source/common/upstream/health_discovery_service.cc

```diff
@@ -298,22 +374,159 @@ HdsCluster::HdsCluster(Server::Admin& admin, Runtime::Loader& runtime,
     hosts_by_locality.back().reserve(locality_endpoints.lb_endpoints_size());
 
     for (const auto& host : locality_endpoints.lb_endpoints()) {
+      const LocalityEndpointTuple endpoint_key = {locality_endpoints.locality(), host};
       // Initialize an endpoint host object.
       HostSharedPtr endpoint = std::make_shared<HostImpl>(
           info_, "", Network::Address::resolveProtoAddress(host.endpoint().address()), nullptr, 1,
           locality_endpoints.locality(),
           envoy::config::endpoint::v3::Endpoint::HealthCheckConfig().default_instance(), 0,
           envoy::config::core::v3::UNKNOWN);
       // Add this host/endpoint pointer to our flat list of endpoints for health checking.
-      initial_hosts_->push_back(endpoint);
+      hosts_->push_back(endpoint);
       // Add this host/endpoint pointer to our structured list by locality so results can be
       // requested by locality.
       hosts_by_locality.back().push_back(endpoint);
+      // Add this host/endpoint pointer to our map so we can rebuild this later.
+      hosts_map_.insert({endpoint_key, endpoint});
     }
   }
   // Create the HostsPerLocality.
-  initial_hosts_per_locality_ =
+  hosts_per_locality_ =
+      std::make_shared<Envoy::Upstream::HostsPerLocalityImpl>(std::move(hosts_by_locality), false);
+}
+
+void HdsCluster::update(Server::Admin& admin, envoy::config::cluster::v3::Cluster cluster,
+                        ClusterInfoFactory& info_factory, ClusterManager& cm,
+                        const LocalInfo::LocalInfo& local_info, Event::Dispatcher& dispatcher,
+                        Random::RandomGenerator& random, Singleton::Manager& singleton_manager,
+                        ThreadLocal::SlotAllocator& tls,
+                        ProtobufMessage::ValidationVisitor& validation_visitor, Api::Api& api,
+                        AccessLog::AccessLogManager& access_log_manager, Runtime::Loader& runtime) {
+  cluster_ = std::move(cluster);
+
+  // Always update our info_.
+  info_ = info_factory.createClusterInfo(
```

Comment:
Will this interact with any ongoing HCs to the unchanged hosts?

Reply:
You are right, I forgot that info_ controls the transport_socket_matches, which is important if a health check changes and needs to use something like TLS. This is the only thing in HDS that is set in v3::Cluster that is used by Host in info_. So now, if the transport_socket_matches change, only then is info_ updated and each host new or old will get an updated copy of info_ so it can use the updated TransportSocketMatcher.